validator/sawtooth_validator/server/state_verifier.py

```diff
@@ -0,0 +1,287 @@
+        max_future_callback_workers=10)
+
+    context_manager = ContextManager(global_state_db)
+
+    transaction_executor = TransactionExecutor(
+        service=component_service,
+        context_manager=context_manager,
+        settings_view_factory=SettingsViewFactory(state_view_factory),
+        scheduler_type=scheduler_type,
+        invalid_observers=[])
+
+    component_service.set_check_connections(
+        transaction_executor.check_connections)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_RECEIPT_ADD_DATA_REQUEST,
+        tp_state_handlers.TpReceiptAddDataHandler(context_manager),
+        component_thread_pool)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_EVENT_ADD_REQUEST,
+        tp_state_handlers.TpEventAddHandler(context_manager),
+        component_thread_pool)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_STATE_DELETE_REQUEST,
+        tp_state_handlers.TpStateDeleteHandler(context_manager),
+        component_thread_pool)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_STATE_GET_REQUEST,
+        tp_state_handlers.TpStateGetHandler(context_manager),
+        component_thread_pool)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_STATE_SET_REQUEST,
+        tp_state_handlers.TpStateSetHandler(context_manager),
+        component_thread_pool)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_REGISTER_REQUEST,
+        processor_handlers.ProcessorRegisterHandler(
+            transaction_executor.processors),
+        component_thread_pool)
+
+    component_dispatcher.add_handler(
+        validator_pb2.Message.TP_UNREGISTER_REQUEST,
+        processor_handlers.ProcessorUnRegisterHandler(
+            transaction_executor.processors),
+        component_thread_pool)
+
+    component_dispatcher.start()
+    component_service.start()
+
+    process_blocks(
+        blocks=blockstore.get_block_iter(start_block=start, reverse=False),
+        transaction_executor=transaction_executor,
+        context_manager=context_manager,
+        state_view_factory=state_view_factory)
+
+    component_dispatcher.stop()
+    component_service.stop()
+    component_thread_pool.shutdown(wait=True)
+
+
+def binary_search_for_state_root(blockstore, state_view_factory):
+    # If there is no chain to process, then we are done.
+    block = blockstore.chain_head
+    if block is None:
+        return None
+
+    # Check the head first
+    if state_db_has_root(state_view_factory, block.state_root_hash):
+        return None
+
+    # If there is a head, do a binary search to find the last block with state
+    max_num = block.block_num
+    min_num = 0
+    next_num = (max_num - min_num) // 2
+
+    while max_num != min_num:
+        block = blockstore.get_block_by_number(next_num)
+
+        # If this block's state root is in state, assume all the lower blocks
```

Comment:
Is this a safe assumption? Have you timed sequential lookup of each block's state root hash from the state database?